File: end2end-example/go/PriceBet.runar.go
package contract

import runar "github.com/icellan/runar/packages/runar-go"

type PriceBet struct {
	runar.SmartContract
	AlicePubKey  runar.PubKey      `runar:"readonly"`
	BobPubKey    runar.PubKey      `runar:"readonly"`
	OraclePubKey runar.RabinPubKey `runar:"readonly"`
	StrikePrice  runar.Bigint      `runar:"readonly"`
}

func (c *PriceBet) Settle(price runar.Bigint, rabinSig runar.RabinSig, padding runar.ByteString, aliceSig runar.Sig, bobSig runar.Sig) {
	msg := runar.Num2Bin(price, 8)
	runar.Assert(runar.VerifyRabinSig(msg, rabinSig, padding, c.OraclePubKey))

	runar.Assert(price > 0)

	if price > c.StrikePrice {
		// bobSig is present in the unlocking script for stack alignment but is
		// intentionally not checked in this branch — only alice (the winner) signs.
		runar.Assert(runar.CheckSig(aliceSig, c.AlicePubKey))
	} else {
		// aliceSig is present in the unlocking script for stack alignment but is
		// intentionally not checked in this branch — only bob (the winner) signs.
		runar.Assert(runar.CheckSig(bobSig, c.BobPubKey))
	}
}

func (c *PriceBet) Cancel(aliceSig runar.Sig, bobSig runar.Sig) {
	runar.Assert(runar.CheckSig(aliceSig, c.AlicePubKey))
	runar.Assert(runar.CheckSig(bobSig, c.BobPubKey))
}

File: end2end-example/rust/PriceBet.runar.rs
use runar::prelude::*;

#[runar::contract]
pub struct PriceBet {
    #[readonly]
    pub alice_pub_key: PubKey,
    #[readonly]
    pub bob_pub_key: PubKey,
    #[readonly]
    pub oracle_pub_key: RabinPubKey,
    #[readonly]
    pub strike_price: Bigint,
}

#[runar::methods(PriceBet)]
impl PriceBet {
    #[public]
    pub fn settle(&self, price: Bigint, rabin_sig: &RabinSig, padding: &ByteString, alice_sig: &Sig, bob_sig: &Sig) {
        let msg = num2bin(&price, 8);
        assert!(verify_rabin_sig(&msg, rabin_sig, padding, &self.oracle_pub_key));

        assert!(price > 0);

        if price > self.strike_price {
            // bob_sig is present in the unlocking script for stack alignment but is
            // intentionally not checked in this branch — only alice (the winner) signs.
            assert!(check_sig(alice_sig, &self.alice_pub_key));
        } else {
            // alice_sig is present in the unlocking script for stack alignment but is
            // intentionally not checked in this branch — only bob (the winner) signs.
            assert!(check_sig(bob_sig, &self.bob_pub_key));
        }
    }

    #[public]
    pub fn cancel(&self, alice_sig: &Sig, bob_sig: &Sig) {
        assert!(check_sig(alice_sig, &self.alice_pub_key));
        assert!(check_sig(bob_sig, &self.bob_pub_key));
    }
}
